calico/felix/ipsets.py

```diff
@@ -383,7 +435,7 @@ def _remove_mapping(self, tag_id, profile_id, endpoint_id, ip_address):
             if not self.ip_owners_by_tag[tag_id]:
                 _log.debug("Tag %s now empty, removing", tag_id)
                 del self.ip_owners_by_tag[tag_id]
-            self._dirty_tags.add(tag_id)
+            self._on_ip_removed(ip_address, tag_id)
```

Comment:
I can see some whacky error conditions where we remove the wrong IP from the _ip_owners_by_tag. It feels like an assert or two in this code would make it more robust (even though you have not changed it, rereading it has made me notice!).